More significant rework of the "multiple handlers" section (#5392)
Particular/docs.particular.net: nservicebus/handlers/index.md

## Invocation of multiple matching handlers

> It is important to keep in mind that, by default, a single incoming message will be processed as a single unit of work, regardless of the number of handlers which are hosted in that endpoint. If one of the handlers fails, the incoming message will be retried according to the [recoverability policy of the endpoint](/nservicebus/recoverability). When the incoming message is retried, _all_ matching handlers get invoked again, including the handlers that had successfully processed the message in previous attempts.

Author: I don't understand. The word "should" here would undermine the "by default" earlier in the sentence.

> It is important to keep in mind that, by default, a single incoming message will be processed as a single unit of work, regardless of the number of handlers which are hosted in that endpoint. If one of the handlers fails, the incoming message will be retried according to the [recoverability policy of the endpoint](/nservicebus/recoverability). When the incoming message is retried, _all_ matching handlers get invoked again, including the handlers that had successfully processed the message in previous attempts.

Reviewer: @udidahan it could be that its not processed as an atomic unit-of-work. To me 'will' feels incorrect.

So maybe it should be the following:
```suggestion
It is important to keep in mind that a single incoming message should be processed as a single unit of work, regardless of the number of handlers which are hosted in that endpoint. If one of the handlers fails, the incoming message will be retried according to the [recoverability policy of the endpoint](/nservicebus/recoverability). When the incoming message is retried, _all_ matching handlers get invoked again, including the handlers that had successfully processed the message in previous attempts.
```

For these reasons, handlers should be designed in such a way that all their operations either rollback if any of them fail, or are idempotent such that they correctly deal with multiple invocations without any side-effects.

If this is not possible, the overall design should be changed such that instead of there being just one message, multiple messages are created. Note that so long as all the handlers continue to be hosted in the same endpoint, these different message will need to be of different types such that each message only matches with one handler. This can be done using one or more of the following techniques:

> - Split the message at the target - intercept the original message with a [pipeline behavior](/nservicebus/pipeline/manipulate-with-behaviors) in which multiple `SendLocal` calls are issued, resulting in multiple messages being processed by the same endpoint. 

Reviewer: Why via a pipeline behavior? Isn't that a lot of overhead as these need to be explicitly registered.

Author: This recommendation is grounded in not making changes to the handler logic. It is my understanding that pipeline behaviors are the documented way of doing that.

Reviewer: It would be a additional handler with the single task to do a few `SendLocal`. Resolving this in a behavior feels unnecessarily complex due to missing auto registration.

Author: Updated the text.

- Split the message at the source - instead of emitting a single message, send out multiple messages directly.

An additional benefit of having multiple messages rather than just one is that the messages can be processed in parallel where otherwise, the execution of the handlers for the single message would have been sequential. 

> Another option that can be considered after creating the multiple messages is to host each handler in its own endpoint. This will provide the greatest degree of isolation allowing independent customization of retry policies, greater visibility, better monitoring, and separate scaling, among other benefits.

Author: While that would work if the message happened to be an event being published, it wouldn't work for commands that are sent to a specific endpoint. 

I suppose that could be called out explicitly. Let me try doing that.

> Another option that can be considered after creating the multiple messages is to host each handler in its own endpoint. This will provide the greatest degree of isolation allowing independent customization of retry policies, greater visibility, better monitoring, and separate scaling, among other benefits.

Author: Done.
